### PR TITLE
[inductor][fx passes] batch layernom

### PR DESCRIPTION
**Pull request by jackiexu1992**
Summary: Batch layernorm. Fuse independent horizontal layernorm with same size into one.

Test Plan:
# unit test
```
buck test mode/dev-nosan //caffe2/test/inductor:group_batch_fusion
File changed: fbcode//caffe2/test/inductor/test_group_batch_fusion.py
Buck UI: https://www.internalfb.com/buck2/68eb51e1-bdbc-4847-aabf-e50737d8485b
Test UI: https://www.internalfb.com/intern/testinfra/testrun/5066549764442206
Network: Up: 0 B  Down: 0 B
Jobs completed: 10. Time elapsed: 1:07.2s.
Tests finished: Pass 3. Fail 0. Fatal 0. Skip 0. Build failure 0
```

Differential Revision: D47447542

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by jackiexu1992** [comment]
@jansel @yanboliang for review.

**Comment by jackiexu1992** [comment]
easycla

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 1]
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
...
+    """
+
+    def match(self, node):
+        if CallFunctionVarArgs(torch.nn.functional.layer_norm).match(node):
+            input_node = get_arg_value(node, 0, "input")
+            group_key = (
+                "batch_layernorm",
+                str(input_node.meta["example_value"].shape)
+                if "example_value" in input_node.meta
+                else "",
```
Comment (jansel) [inline torch/_inductor/fx_passes/group_batch_fusion.py:171]: What happens with missing example values?  Should this prevent fusion?

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 2]
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
...
+        group_weights = []
+        group_biases = []
+        group_epss = []
+        group_nodes = []
+        for node in subset:
+            group_nodes.append(node)
+            group_inputs.append(get_arg_value(node, 0, "input"))
+            group_shapes.append(get_arg_value(node, 1, "normalized_shape"))
+            group_weights.append(get_arg_value(node, 2, "weight"))
+            group_biases.append(get_arg_value(node, 3, "bias"))
```
Comment (jansel) [inline torch/_inductor/fx_passes/group_batch_fusion.py:191]: `bias` is an optional arg.  What happens if some have bias set and others don't?

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 3]
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
...
+        group_shapes = []
+        group_weights = []
+        group_biases = []
+        group_epss = []
+        group_nodes = []
+        for node in subset:
+            group_nodes.append(node)
+            group_inputs.append(get_arg_value(node, 0, "input"))
+            group_shapes.append(get_arg_value(node, 1, "normalized_shape"))
+            group_weights.append(get_arg_value(node, 2, "weight"))
```
Comment (jansel) [inline torch/_inductor/fx_passes/group_batch_fusion.py:190]: `weight` is an optional arg.   What happens if some have weight set and others don't?

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 4]
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
...
+            group_biases.append(get_arg_value(node, 3, "bias"))
+            eps = get_arg_value(node, 4, "eps")
+            if eps is None:
+                eps = 1e-5
+            group_epss.append(eps)
+        stack_dim = -1 - len(group_shapes[-1])
```
Comment (jansel) [inline torch/_inductor/fx_passes/group_batch_fusion.py:196]: Add asserts that each item in these lists matches.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/group_batch_fusion.py (hunk 1 above)
Comment (jackiexu1992) [reply torch/_inductor/fx_passes/group_batch_fusion.py:171]: Make sense. Should prevent such fusions.

**Comment by jackiexu1992** [comment]
Comments resolved :) cc @jansel

**Review by jansel — APPROVED** [review]

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 5]
```diff
@@ -156,6 +156,116 @@ def fuse(self, graph, subset):
...
+            weight = get_arg_value(node, 2, "weight")
+            bias = get_arg_value(node, 3, "bias")
+            group_key = (
+                (
+                    "batch_layernorm",
+                    str(input.meta["example_value"].shape),
+                    str(weight.meta["example_value"].shape) if weight is not None else "",
+                    str(get_arg_value(node, 1, "normalized_shape")),
+                    weight is None,
+                    bias is None,
```
Comment (jansel) [inline torch/_inductor/fx_passes/group_batch_fusion.py:184]: Do these need to be shapes too?  

Not sure if things like broadcasting are legal.

**Review by xuzhao9 — APPROVED** [review]

**Pull request closed without merging**